### PR TITLE
nv_peer_mem: Bar address re-mapping for POWER

### PR DESCRIPTION
**Pull request by ferasd**
Support address remapping on Power systems and future Intel systems
by using Nvidia API to map physical pages retrieved using
nvidia_p2p_get_pages accessible to the HCA. The change is backward
compatible that works on older systems.

Signed-off-by: Feras Daoud <ferasda@mellanox.com>

### CODE REVIEW

File: nv_peer_mem.c [hunk 1]
```diff
@@ -39,18 +39,67 @@
...
+#include <linux/pci.h>
 
 
 #include "nv-p2p.h"
 #include <rdma/peer_mem.h>
 
 
 #define DRV_NAME	"nv_mem"
-#define DRV_VERSION	"1.0"
+#define DRV_VERSION	"1.1"
```
Comment (haggaie) [inline nv_peer_mem.c]: I would recommend putting version changes in a separate patch.

File: nv_peer_mem.c [hunk 2]
```diff
@@ -39,18 +39,67 @@
...
 #define DRV_RELDATE	__DATE__
 
-#define peer_err(FMT, ARGS...) printk(KERN_ERR   DRV_NAME " %s:%d " FMT, __FUNCTION__, __LINE__, ## ARGS)
+#define peer_err(FMT, ARGS...) printk(KERN_ERR   DRV_NAME " %s:%d " FMT, __func__, __LINE__, ## ARGS)
```
Comment (haggaie) [inline nv_peer_mem.c]: This also doesn't seem to be related to the patch.

File: nv_peer_mem.c [hunk 3]
```diff
@@ -39,18 +39,67 @@
...
 
+#define INFO_ENABLE 1
+#define peer_info(FMT, ARGS...)		\
```
Comment (haggaie) [inline nv_peer_mem.c]: This is not used by the patch, right? Also, if you want programmable prints, I recommend using [dynamic-debug](https://www.kernel.org/doc/Documentation/dynamic-debug-howto.txt) if it is compiled in the kernel.

File: nv_peer_mem.c [hunk 4]
```diff
@@ -39,18 +39,67 @@
...
+	(((v) & NVIDIA_P2P_MINOR_VERSION_MASK))
+#endif
+
+/*
+ *	Note: before major version 2, struct dma_mapping had no version field,
+ *	so it is not possible to check version compatibility. In this case
+ *	let us just avoid dma mappings altogether.
+ */
+#if defined(NVIDIA_P2P_DMA_MAPPING_VERSION) &&	\
+	(NVIDIA_P2P_MAJOR_VERSION(NVIDIA_P2P_DMA_MAPPING_VERSION) >= 2)
```
Comment (haggaie) [inline nv_peer_mem.c]: Who defines `NVIDIA_P2P_DMA_MAPPING_VERSION`? Is it in the nv-p2p.h header file?

File: nv_peer_mem.c [hunk 5]
```diff
@@ -39,18 +39,67 @@
...
+#pragma message("Enable nvidia_p2p_dma_map_pages support")
+#define NV_DMA_MAPPING 1
+#else
+#define NV_DMA_MAPPING 0
+#endif
+
+#ifndef NVIDIA_P2P_MAJOR_VERSION_MATCHES
+#define NVIDIA_P2P_MAJOR_VERSION_MATCHES(p, v)	\
```
Comment (haggaie) [inline nv_peer_mem.c]: This is actually code that runs on run-time, right? Why is it a macro?

File: nv_peer_mem.c [hunk 6]
```diff
@@ -187,18 +239,58 @@ static int nv_dma_map(struct sg_table *sg_head, void *context,
...
+		if (!NVIDIA_P2P_DMA_MAPPING_VERSION_COMPATIBLE(dma_mapping)) {
+			peer_err("error, incompatible dma mapping version 0x%08x\n",
+				 dma_mapping->version);
+			nvidia_p2p_dma_unmap_pages(pdev, page_table, dma_mapping);
+			return -EINVAL;
+		}
+
+		nv_mem_context->npages = dma_mapping->entries;
+
+		ret = sg_alloc_table(sg_head, dma_mapping->entries, GFP_KERNEL);
```
Comment (haggaie) [inline nv_peer_mem.c]: Who releases the table? Is it the caller or the unmap function?

File: nv_peer_mem.c (hunk 1 above)
Comment (ferasd) [inline nv_peer_mem.c]: moved to new commit

File: nv_peer_mem.c (hunk 2 above)
Comment (ferasd) [inline nv_peer_mem.c]: reverting the change

File: nv_peer_mem.c (hunk 3 above)
Comment (ferasd) [inline nv_peer_mem.c]: good idea, removing the print

File: nv_peer_mem.c (hunk 4 above)
Comment (ferasd) [inline nv_peer_mem.c]: yes

File: nv_peer_mem.c (hunk 5 above)
Comment (ferasd) [inline nv_peer_mem.c]: unused code, removed

File: nv_peer_mem.c (hunk 6 above)
Comment (ferasd) [inline nv_peer_mem.c]: On success, nv_mem_put_pages frees the memory otherwise the sg_alloc_table frees any leftovers

**Comment by ferasd** [comment]
@haggaie
